### PR TITLE
add tcdp resource code from old pr

### PR DESCRIPTION
**Pull request by xh3b4sd**
Successor of https://github.com/giantswarm/aws-operator/pull/1488.

### CODE REVIEW

File: service/controller/clusterapi/v29/key/common.go [hunk 1]
```diff
@@ -31,6 +31,10 @@ func IsDeleted(getter DeletionTimestampGetter) bool {
 	return getter.GetDeletionTimestamp() != nil
 }
 
+func MachineDeploymentASGName(getter LabelsGetter) string {
+	return fmt.Sprintf("asg-%s-tcnp", ClusterID(getter))
```
Comment (xh3b4sd) [inline service/controller/clusterapi/v29/key/common.go]: That is just made up with the new naming scheme.

File: service/controller/clusterapi/v29/resource/tcdp/create.go [hunk 2]
```diff
@@ -2,22 +2,29 @@ package tcdp
...
+	"github.com/giantswarm/aws-operator/service/controller/clusterapi/v29/resource/tcdp/template"
 )
 
 const (
 	capabilityNamesIAM = "CAPABILITY_NAMED_IAM"
 )
 
 func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
-	cr, err := key.ToMachineDeployment(obj)
+	cl, err := r.toClusterFunc(obj)
```
Comment (xh3b4sd) [inline service/controller/clusterapi/v29/resource/tcdp/create.go]: I find it weird that the cluster type is necessary here but it does not easily work without it now.

File: service/controller/clusterapi/v29/resource/tcdp/create.go [hunk 3]
```diff
@@ -120,22 +156,200 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
...
+		Cluster: template.ParamsMainAutoScalingGroupCluster{
+			ID: key.ClusterID(&md),
+		},
+		DesiredCapacity:       minDesiredNodes,
+		MaxBatchSize:          workerCountRatio(minDesiredNodes, 0.3),
+		MaxSize:               key.WorkerScalingMax(md),
+		MinInstancesInService: workerCountRatio(minDesiredNodes, 0.7),
+		MinSize:               key.WorkerScalingMin(md),
+		Name:                  key.MachineDeploymentASGName(&md),
+		Subnets:               nil, // TODO
```
Comment (xh3b4sd) [inline service/controller/clusterapi/v29/resource/tcdp/create.go]: There are several TODOs here now which we need to address in upcoming PRs. I am happy for now to got this mess compiled. 🙈 hashtagbabysteps

File: service/controller/clusterapi/v29/resource/tcdp/create.go [hunk 4]
```diff
@@ -120,22 +156,200 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
...
+// might be that initial ASG configuration is following:
+// 		- Min: 3
+//		- Max: 10
+// 		- Desired: 10
+//
+// Now end user decides that it must be scaled down so maximum size is decreased
+// to 7. When desired number of instances is temporarily bigger than maximum
+// number of instances, it must be fixed to be maximum number of instances.
+//
+func minDesiredWorkers(minWorkers, maxWorkers, statusDesiredCapacity int) int {
```
Comment (xh3b4sd) [inline service/controller/clusterapi/v29/resource/tcdp/create.go]: This is copied from the adapter package. Our goal is to get rid of it over time so I think this is the better place now.

File: service/controller/clusterapi/v29/key/common.go (hunk 1 above)
Comment (tuommaki) [inline service/controller/clusterapi/v29/key/common.go]:
```suggestion
	return fmt.Sprintf("asg-%s-tcnp-%s", ClusterID(getter), MachineDeploymentID(getter))
```
No strong feelings about the string in general, but I thought it makes sense to identify it by cluster and machine deployment (as there will be several machine deployments per cluster).

File: service/controller/clusterapi/v29/resource/tcdp/create.go (hunk 2 above)
Comment (tuommaki) [reply service/controller/clusterapi/v29/resource/tcdp/create.go]: Sure, we can improve things like this with iteration.

File: service/controller/clusterapi/v29/key/common.go (hunk 1 above)
Comment (xh3b4sd) [reply service/controller/clusterapi/v29/key/common.go]: Good point. I added some version of this.

**Pull request merged**